programmer-skills/big-O-notation.md

```diff
@@ -0,0 +1,161 @@
+> |f(x)|&le;M|g(x)| for all x&ge; x<sub>0</sub>.
+ 
+ 
+Let's discuss the big-O notation through several examples
+
+###O(1)
+                                                          
+O(1) represents an algorithm that has the same runtime regardless of the size of the input data set.
+
+An easy example is when we check whether the first element of the array is null. The algorithm always checks 
+the first element in the array, regardless of the array's size.
+
+```javascript
+function isFirstItemNull(array)
+{
+	return array[0] === null;
+}
+```
+
+###O(N)
+
+O(N) describes an algorithm whose runtime will grow linearly and in direct proportion to the size of the input data set. For example  if the algorithm includes a single iteration
+ over the data set, the perfromance is O(N). 
+
+Our test algorithm checks whether an item is in the array or not.
+```javascript
+function containsItem(array, item){
+	for (var element in array)
+	{
+		if (item === array[element]) return true;
+	}
+	return false;
+}
+```
+
+###O(N<sup>2</sup>)
+The performance is O(N<sup>2</sup>) when the runtime is directly proportional to the square of the size of the input data set.
+Almost every algorithm which includes a nested iteration over the data set has O(N<sup>2</sup>) performance.
+E.g.: bubble sort (worst case), Shell sort, quicksort (worst case), selection sort or insertion sort.
+
+The everyCombination function lists every possible ordered pair made from the elements of an array.
+It iterates through the array and for every element (item) it lists all of the ordered pairs starting with (item), using an inner loop. 
+
+Sample output: If our array is [1,2], then the returned array containing all the possible combinations is [[1,1],[1,2],[2,1],[2,2]]. 
+ 
+```javascript
+function everyCombination(array)
+{
+	var result = [];
+	for (var item in array)
+	{
+		for (var inner in array)
+		{
+			result.push( [array[item], array[inner]] )
+		}
+	}
+	return result;
+}
+```
+Another example for O(N<sup>2</sup>) algorithms is when we check whether all items are unique in an array. The number of comparisons is 
+(N-1) + (N-2) + (N-3)...+ 2+1 = N(N-1)/2.
+
+```javascript
+function isUnique(array)
+{
+	for (var i = 0; i < array.length-1; i++)
+	{
+		for (var j = i+1; j< array.length; j++)
+		{
+			if (array[i]==array[j] ) return false;
+		}
+	}
+	return true;
+}
+```
+
+###O(2<sup>N</sup>)
+O(2<sup>N</sup>) corresponds to an algorithm where adding a  new item to the data set doubles the number of steps.
+In case of O(2<sup>N</sup>) algrithms the number of steps grow exponentially.
+
+A well-known example is the naive recursive calculation of the Fibonnaci-series.
+In general:
+f(n) = f(n-1) + f(n-2) = f(n-2) + f(n-3) + f(n-3) + f(n-4) =
+= f(n-3) + f(n-4) + f(n-4) + f(n-5) + f(n-4) + f(n-5) + f(n-5) + f(n-6) = ...
+
+```javascript
+	function fibonacci(n)
+	{
+		if (n===0 || n===1) return 1;
+		else return ( fibonacci(n-1) + fibonacci(n-2) );
+	}
+```
+
+###O(log N)
```

Comment:
Would it make sense to move this algorithm up in the document, below O(1)? In this way you'd have the algorithms listed in order of increasing runtime, and also in the same order as they are in the figure.